src/java/org/openzal/zal/Acl.java

```diff
@@ -86,14 +87,23 @@ public boolean isEmpty()
   public Grant grantAccess(String zimbraId, byte type, short rights, String secret)
     throws ZimbraException
   {
-    ACL.Grant grant;
-    try
-    {
+    ACL.Grant grant = null;
+    try {
       grant = mAcl.grantAccess(zimbraId, type, rights, secret);
-    }
-    catch (com.zimbra.common.service.ServiceException e)
-    {
-      throw ExceptionWrapper.wrap(e);
+    } catch (com.zimbra.common.service.ServiceException e) {
+      if (!e.getCode().equals(MailServiceException.GRANT_EXISTS)) {
+        throw ExceptionWrapper.wrap(e);
+      }
+
+      List<ACL.Grant> grants = mAcl.getGrants();
+      for (ACL.Grant g : grants) {
+        //getting the grant even in case a GRANT_EXISTS exception has been thrown
+        if (g.isGrantee(zimbraId) && g.getGrantedRights() == rights &&
+            ((type != GRANTEE_GUEST && type != GRANTEE_KEY) ||
+                StringUtil.equal(g.getPassword(), secret))) {
+          grant = g;
+        }
+      }
```

Comment:
you can stop the loop after you find the grant, if gran == null after the loop you should throw ExceptionWrapper.wrap(e)